src/systems/perception.rs

```diff
@@ -0,0 +1,70 @@
+use amethyst::{
+    core::{nalgebra::Vector4, transform::GlobalTransform},
+    ecs::{
+        Entities, Join, Read, ReadExpect, ReadStorage, System, Write, WriteExpect, WriteStorage,
+    },
+    renderer::DebugLines,
+};
+
+use crate::components::perception::Perception;
+use crate::resources::{debug::DebugConfig, spatial_grid::SpatialGrid};
+
+pub struct EntityDetectionSystem;
+
+impl<'s> System<'s> for EntityDetectionSystem {
+    type SystemData = (
+        Entities<'s>,
+        WriteStorage<'s, Perception>,
+        ReadExpect<'s, SpatialGrid>,
+        ReadStorage<'s, GlobalTransform>,
+        Write<'s, DebugLines>,
+        Read<'s, DebugConfig>,
+    );
+
+    fn run(
+        &mut self,
+        (entities, mut perceptions, grid, globals, mut debug_lines, debug_config): Self::SystemData,
+    ) {
+        for (entity, mut perception, global) in (&entities, &mut perceptions, &globals).join() {
+            perception.entities = Vec::new();
+            let nearby_entities = grid.query(global, perception.range);
+            let pos = Vector4::from(global.as_ref()[3]).xyz();
+            let sq_range = perception.range * perception.range;
+
+            for other_entity in nearby_entities {
+                if entity == other_entity {
+                    continue;
+                }
+                let other_global = globals.get(other_entity).unwrap();
```

Comment:
This might fail if `other_entity` is dead by the time we run this system

Reply:
I changed the loop to use a `BitSet` joined with `Entities` to make sure we're iterating over valid entities at that point.